### DISCUSSION TITLE
Missing code or SDK

### DISCUSSION THREAD

**Discussion by Imar** [dojo-engineering/feedback#4] [Docs]
Hi there,

On pages like this: https://docs.dojo.tech/docs/development-resources/webhooks#step-3-verify-dojo-webhooks you show sample code that uses classes like \WebhookPayloadUtils. However, these are nowhere to be found, making it pretty difficult to implement signatue validation.

Can the samples be updated with working code or can the code used in the examples be made available online?

Thanks!

Imar

**Comment by martinhay1** [comment]
Hi Imar,

Thanks for reaching out. I believe this is what you might be looking for: 
https://github.com/dojo-engineering/Dojo.Net/blob/main/src/Webhooks/WebhookPayloadUtils.cs

Please let us know if you have further queries, thanks.

Martin Hay
Dojo Support

**Reply by Imar** [reply]
Hi Martin,

Thanks for that; that solves some of the problems (in particular the issue with WebhookPayloadUtils). However, other types, like WebhooksClient are still missing.

It looks like that most of what's described here: https://github.com/dojo-engineering/Dojo.Net is either missing or invalid. The list of .NET Frameworks for the Dojo SDK is incorrect. For example, you can't install it into netstandard2.0 or ,NET Framework 4.x. So, if you want to build code that targets these frameworks, you have to build it yourself. It would be great if either: a) the Dojo.Net library would support the frameworks it advertises, or that the full code is available so we can easily build it ourselves for other frameworks without having to reinvent the wheel and open up for potential security issues.

Thanks!

Imar

**Reply by RomanKondratyev** [reply]
Hi @Imar ,

I already answered you in the other thread - I will update this to support netstandard2.0 and let you know - should be pretty straightforward.

Thanks,
Roman

**Reply by Imar** [reply]
Thanks for that. Will you also consider publishing the source code?

**Reply by RomanKondratyev** [reply]
@Imar no problems. The code is 100% there already. In order to get the `clients` you need to build the project, using .NET60+. I will update the readme to reflect this, but in general:
`cd src`
`dotnet build`

**Reply by Imar** [reply]
Oh, I see now. It uses NSwag.MSBuild to build the clients dynamically. That was the part I was missing, so yes, it would be nice if the readme made that more clear.
Thanls!
Imar

**Reply by RomanKondratyev** [reply]
@Imar the issue is closed. Please see here: https://github.com/dojo-engineering/Dojo.Net/issues/1

**Reply by gyroscopes** [reply]
I've been really struggling with the C# examples. Put in a few calls about it but have no help/suggestions. Code has just not been compiling. I didn't know if there was code missing or some critical information such as a particular configuration/version of visual studio.

Can some update the documentation because it is very misleading and dojo technical support hasn't responded to my questions.

**Reply by RomanKondratyev** [reply]
Hi @gyroscopes,

Thank you for your feedback. Could you be more specific which example is not compiling - then we can fix all the issues.

**Reply by RomanKondratyev** [reply]
On the support note, please feel free to ask questions here instead of contacting support.

**Reply by gyroscopes** [reply]
webhooks/cs/verify-webhooks.cs is a good example that caused me confusion.

The code shown is in the example
```
// read request body as a byte stream
using var streamReader = new StreamReader(HttpContext.Request.Body);
var body = await streamReader.ReadToEndAsync();

// Get signature header value
if (HttpContext.Request.Headers.TryGetValue(WebhookPayloadUtils.SignatureHeaderName, out var signatureHeader))
{
    // use Dojo.Net SDK to deserialize and validate webhook payload
    var payload = WebhookPayloadUtils.ReadPayload(body, _webhookSecret, signatureHeader);

    // TODO: update your database or do some action based on successful payment
}

```

I've spent hours wondering what version of C# language this is and what configuration of visual studio it is.
It won't compile without more code and without knowing a critical bit of information.

Firstly this is the full code
```
using System;
using System.Collections.Generic;
using System.Diagnostics.CodeAnalysis;
using System.IO;
using System.Threading;
using System.Threading.Tasks;
using Microsoft.AspNetCore.Mvc;
using Dojo.Net;
using Microsoft.Extensions.Logging;
using ProblemDetails = Dojo.Net.ProblemDetails;
using Dojo.Net.Webhooks;
using Microsoft.Extensions.Configuration;
using server.Model;

namespace Server.Controllers;

[ApiController]
[Route("[controller]")]
public class WebhookController : ControllerBase
{
    private readonly ILogger<WebhookController> _logger;
    private readonly IPaymentIntentsClient _paymentIntentsClient;
    private readonly string _webhookSecret;

    public WebhookController(IConfiguration configuration, ILogger<WebhookController> logger, IPaymentIntentsClient paymentIntentsClient)
    {
        // do not store secrets as plain text in the config files, consider to use secret managers
        _webhookSecret = configuration.GetSection("Dojo")["WebhookSecret"];

        _logger = logger ?? throw new ArgumentNullException(nameof(logger));
        _paymentIntentsClient = paymentIntentsClient ?? throw new ArgumentNullException(nameof(paymentIntentsClient));
    }

    [HttpPost]
    public async Task WebhookCallback(CancellationToken cancellationToken)
    {
        using var streamReader = new StreamReader(HttpContext.Request.Body);
        var body = await streamReader.ReadToEndAsync();

        if (HttpContext.Request.Headers.TryGetValue(WebhookPayloadUtils.SignatureHeaderName, out var signatureHeader))
        {
            var payload = WebhookPayloadUtils.ReadPayload(body, _webhookSecret, signatureHeader);

            if (payload.Data.PaymentStatus == PaymentIntentStatus.Captured)
            {
                var paymentIntent = await _paymentIntentsClient.GetAsync(payload.Data.PaymentIntentId, cancellationToken);
                // TODO: update your database or do some action based on successful payment
                _logger.LogInformation($"Payment successfull for paymentIntentId: {payload.Data?.PaymentIntentId}");
            }

            _logger.LogInformation($"Received event. PaymentIntentId: {payload.Data?.PaymentIntentId}, Status: {payload.Data?.PaymentStatus}");
        }
    }
}
```

Secondly (the critical part) it needs to run in the controller part of MVP project. This probably obvious for many C# programmers that have done lots of new projects. But if you have not seen a MVP project due to working on legacy C# systems it confusing.

Maybe use abbreviated code on developer website but provide links to full code and example its for a MVP project.

**Reply by RomanKondratyev** [reply]
I actually agree, the example could use more generic code (not using MVC specifics) and be more complete. @monica-m-ps
